sway-core/src/asm_generation/fuel/register_allocator.rs

```diff
@@ -385,51 +483,305 @@ pub(crate) fn coalesce_registers(
 ///
 /// Algorithm:
 /// ===============================================================================================
-/// 1. Pick any node n such that degree(n) < k and put it on the stack along with its neighbors.
+/// 1. Pick any node n such that degree(n) < k and put it on the stack.
 /// 2. Remove node n and all its edges from the graph
 ///    - This may make some new nodes have fewer than k neighbours which is nice.
-/// 3. If some vertex n still has k or more neighbors, then the graph is not k colorable, and we
-///    have to spill. For now, we will just error out, but we may be able to do better in the
-///    future.
+/// 3. If some vertex n still has k or more neighbors, then the graph may not be k colorable.
+///     We still add it to the stack as is, as a potential spill. When popping, if we still
+///     can't colour it, then it becomes an actual spill.
 /// ===============================================================================================
 ///
-/// As we don't implement spilling just yet, I've modified the algorithm above to assume k=infinity
-/// and moving the colorability checking until the register assignment phase. The reason for this
-/// is that the algorithm above can be too conservative and may bail our early even though a valid
-/// assignment is actually available. We can revisit this decision when decide to implement
-/// spilling.
-///
 pub(crate) fn color_interference_graph(
     interference_graph: &mut InterferenceGraph,
-) -> Vec<(VirtualRegister, BTreeSet<VirtualRegister>)> {
+    ops: &[Op],
+    live_out: &[FxHashSet<VirtualRegister>],
+) -> Result<Vec<NodeIndex>, FxHashSet<VirtualRegister>> {
     let mut stack = Vec::with_capacity(interference_graph.node_count());
+    let mut on_stack = FxHashSet::default();
+    let mut spills = FxHashSet::default();
+    let def_use_points = compute_def_use_points(ops);
+
+    // for op in ops {
+    //     println!("{}", op);
+    // }
+    // println!("");
 
-    // Raw for loop here is safe because we are not actually removing any nodes from the graph at
-    // any point (i.e. we're never calling `remove_node()`). This means that each `index` below
-    // correspond to a valid `NodeIndex` in the graph.
-    for index in 0..interference_graph.node_count() {
-        // Convert to a `NodeIndex`
-        let node = node_index(index);
+    // for node_index in interference_graph.node_indices() {
+    //     let node = &interference_graph[node_index];
+    //     print!(
+    //         "Neighbours of {} (o{}, i{}): ",
+    //         node,
+    //         interference_graph.neighbors_directed(node_index, Outgoing).count(),
+    //         interference_graph.neighbors_directed(node_index, Incoming).count()
+    //     );
+    //     for neighbour in interference_graph.neighbors(node_index) {
+    //         print!("{} ", interference_graph[neighbour]);
+    //     }
+    //     println!("");
+    // }
```

Comment:
Are these commented out because they will be useful in the future or can they be deleted?

Reply:
I left it for future debugging but it isn't that complex that it can't be rewritten in a short time. Deleting it now.